File: src/tuple.cpp
#include "tuple.h"
#include <cmath>

class Tuple;
class Point;
class Vector;

Tuple::Tuple()
{
	mbrX = 0;
	mbrY = 0;
	mbrZ = 0;
	mbrW = 0;
}
Tuple::Tuple(double argx, double argy, double argz, double argw)
{
	mbrX = argx;
	mbrY = argy;
	mbrZ = argz;
	mbrW = argw;
};
Tuple Tuple::operator-()
{
	return Tuple(-mbrX, -mbrY, -mbrZ, -mbrW);
}

Tuple Tuple::operator*(float multiple)
{
	return Tuple(mbrX * multiple, mbrY * multiple, mbrZ * multiple, mbrW * multiple);
}
Tuple Tuple::operator/(float multiple)
{
	return Tuple(mbrX / multiple, mbrY / multiple, mbrZ / multiple, mbrW / multiple);
}
Tuple Tuple::operator+(Tuple a)
{
	return Tuple(a.mbrX + mbrX, a.mbrY + mbrY, a.mbrZ + mbrZ, a.mbrW + mbrW);;
}
Tuple Tuple::add(Tuple a)
{
	return Tuple(a.mbrX + mbrX, a.mbrY + mbrY, a.mbrZ + mbrZ, a.mbrW + mbrW);
};
Tuple Tuple::subtract(Tuple a)
{
	return Tuple(mbrX - a.mbrX, mbrY - a.mbrY, mbrZ - a.mbrZ, mbrW - a.mbrW);
};
Tuple Tuple::negate()
{
	return Tuple(-mbrX, -mbrY, -mbrZ, -mbrW);
};
float Tuple::magnitude()
{
	return sqrt(pow(mbrX, 2) + pow(mbrY, 2) + pow(mbrZ, 2) + pow(mbrW, 2));
}
double Tuple::dot(Tuple a)
{
	return (mbrX * a.mbrX + mbrY * a.mbrY + mbrZ * a.mbrZ);
};
Tuple Tuple::reflect(Tuple normal)
{
	return subtract(normal) * 2 * dot(normal);
}

Point::Point() : Tuple(0, 0, 0, 1.0) {};
Point::Point(double argx, double argy, double argz) : Tuple(argx, argy, argz, 1.0) {};
Point Point::operator+(Tuple a)
{
	return Point(a.mbrX + mbrX, a.mbrY + mbrY, a.mbrZ + mbrZ);;
}
Vector Point::operator-(Point a)
{
	return Vector(mbrX-a.mbrX, mbrY-a.mbrY, mbrZ-a.mbrZ);;
}
Point Point::subtract(Tuple a)
{
	return Point(mbrX - a.mbrX, mbrY - a.mbrY, mbrZ - a.mbrZ);
}

Vector::Vector() : Tuple(0, 0, 0, 0.0) {};
Vector::Vector(double argx, double argy, double argz) : Tuple(argx, argy, argz, 0.0) {};
Vector Vector::operator+(Tuple a)
{
	return Vector(a.mbrX + mbrX, a.mbrY + mbrY, a.mbrZ + mbrZ);;
}
Vector Vector::operator-()
{
	return Vector(-mbrX, -mbrY, -mbrZ);
}
Vector Vector::operator*(float multiple)
{
	return Vector(mbrX * multiple, mbrY * multiple, mbrZ * multiple);
}
Vector Vector::subtract(Tuple a)
{
	return Vector(mbrX - a.mbrX, mbrY - a.mbrY, mbrZ - a.mbrZ);
}
Vector Vector::normalize()
{
	return Vector(mbrX / magnitude(), mbrY / magnitude(), mbrZ / magnitude());
};
double Vector::dot(Vector a)
{
	return (mbrX * a.mbrX + mbrY * a.mbrY + mbrZ * a.mbrZ);
};
Vector Vector::cross(Vector a)
{
	return Vector(mbrY * a.mbrZ - mbrZ * a.mbrY, mbrZ * a.mbrX - mbrX * a.mbrZ, mbrX * a.mbrY - mbrY * a.mbrX);
};
Vector Vector::reflect(Vector normal)
{
	double varDot = dot(normal);
	varDot = varDot * 2;
	Vector ref = normal * varDot;
	ref = subtract(ref);
	return  ref;
}

File: src/Tuple.cs
using LibComparinator;

namespace LibTuple;
public class SpaceTuple {
	public double _fieldX;
	public double _fieldY;
	public double _fieldZ;
	public double _fieldW;
    public SpaceTuple () {
	    _fieldX = 0.0;
	    _fieldY = 0.0;
	    _fieldZ = 0.0;
	    _fieldW = 0.0;
    }
    public SpaceTuple(double argx, double argy, double argz, double argw) {
        _fieldX = argx;
        _fieldY = argy;
        _fieldZ = argz;
        _fieldW = argw;
    }
    public SpaceTuple(SpaceTuple argOther) {
        _fieldX = argOther._fieldX;
        _fieldY = argOther._fieldY;
        _fieldZ = argOther._fieldZ;
        _fieldW = argOther._fieldW;
    }
    public static SpaceTuple operator-(SpaceTuple argSelf) { return new SpaceTuple(-argSelf._fieldX, -argSelf._fieldY, -argSelf._fieldZ, -argSelf._fieldW); }
    public static SpaceTuple operator-(SpaceTuple argSelf, SpaceTuple argOther) { return new SpaceTuple(argSelf._fieldX-argOther._fieldX, argSelf._fieldY-argOther._fieldY, argSelf._fieldZ-argOther._fieldZ, argSelf._fieldW-argOther._fieldW); }
    public static SpaceTuple operator+(SpaceTuple argSelf, SpaceTuple argOther) { return new SpaceTuple(argSelf._fieldX+argOther._fieldX, argSelf._fieldY+argOther._fieldY, argSelf._fieldZ+argOther._fieldZ, argSelf._fieldW+argOther._fieldW); }
    public static SpaceTuple operator*(SpaceTuple argSelf, double multiple) { return new SpaceTuple(argSelf._fieldX * multiple, argSelf._fieldY * multiple, argSelf._fieldZ * multiple, argSelf._fieldW * multiple); }
    public static SpaceTuple operator/(SpaceTuple argSelf, double multiple) { return new SpaceTuple(argSelf._fieldX / multiple, argSelf._fieldY / multiple, argSelf._fieldZ / multiple, argSelf._fieldW / multiple); }
    public double GetMagnitude() { return Math.Sqrt(_fieldX * _fieldX) + (_fieldY * _fieldY) + (_fieldZ * _fieldZ) + (_fieldW* _fieldW); }
    public double GetDotProduct(SpaceTuple argOther) { return (_fieldX * argOther._fieldX) + (_fieldY * argOther._fieldY) + (_fieldZ * argOther._fieldZ) + (_fieldW * argOther._fieldW); }
    bool CheckEqual(SpaceTuple argOther)
    {
        Comparinator varComp = new Comparinator ();
        return varComp.CheckTuple(this, argOther);
    }
    public void RenderConsole() {
        Console.WriteLine($"Tuple::renderConsole() -> {{mbrX:{_fieldX}, mbrY:{_fieldY}, mbrZ:{_fieldZ}, mbrW:{_fieldW}}}");
    }
    public void SetAdd (SpaceTuple argOther) {
        _fieldX += argOther._fieldX;
        _fieldY += argOther._fieldY;
        _fieldZ += argOther._fieldZ;
        _fieldW += argOther._fieldW;
    }
    public void SetSubtract (SpaceTuple argOther) {
        _fieldX -= argOther._fieldX;
        _fieldY -= argOther._fieldY;
        _fieldZ -= argOther._fieldZ;
        _fieldW -= argOther._fieldW;
    }
    public void SetNegative()
    {
        _fieldX = -_fieldX;
        _fieldY = -_fieldY;
        _fieldZ = -_fieldZ;
        _fieldW = -_fieldW;
    }
}
public class Point : SpaceTuple {
    public Point() {
	    _fieldX = 0.0;
	    _fieldY = 0.0;
	    _fieldZ = 0.0;
	    _fieldW = 1.0;
    }
    public Point(double argX, double argY, double argZ) {
	    _fieldX = argX;
	    _fieldY = argY;
	    _fieldZ = argZ;
	    _fieldW = 1.0;
    }
    public static Vector operator-(Point argSelf, Point argOther) { return new Vector(argSelf._fieldX-argOther._fieldX, argSelf._fieldY-argOther._fieldY, argSelf._fieldZ-argOther._fieldZ); }
    public static Point operator-(Point argSelf, SpaceTuple argOther) { return new Point(argSelf._fieldX-argOther._fieldX, argSelf._fieldY-argOther._fieldY, argSelf._fieldZ-argOther._fieldZ); }
    public static Point operator+(Point argSelf, SpaceTuple argOther) { return new Point(argSelf._fieldX+argOther._fieldX, argSelf._fieldY+argOther._fieldY, argSelf._fieldZ+argOther._fieldZ); }
}
public class Vector : SpaceTuple {
    public Vector() {
	    _fieldX = 0.0;
	    _fieldY = 0.0;
	    _fieldZ = 0.0;
	    _fieldW = 0.0;
    }
    public Vector(double argX, double argY, double argZ) {
	    _fieldX = argX;
	    _fieldY = argY;
	    _fieldZ = argZ;
	    _fieldW = 0.0;
    }
    // Vector Vector::operator+(const Tuple &argOther)
    // {
    // 	return Vector(argOther.mbrX + mbrX, argOther.mbrY + mbrY, argOther.mbrZ + mbrZ);;
    // }
    // Vector Vector::operator-()
    // {
    // 	return Vector(-mbrX, -mbrY, -mbrZ);
    // }
    public static Vector operator*(Vector argSelf, double multiple) {
    	return new Vector(argSelf._fieldX * multiple, argSelf._fieldY * multiple, argSelf._fieldZ * multiple);
    }
    public static Vector operator-(Vector argSelf) { return new Vector(-argSelf._fieldX, -argSelf._fieldY, -argSelf._fieldZ); }
    public static Vector operator-(Vector argSelf, SpaceTuple argOther) { return new Vector(argSelf._fieldX-argOther._fieldX, argSelf._fieldY-argOther._fieldY, argSelf._fieldZ-argOther._fieldZ); }
    public static Vector operator-(Vector argSelf, Vector argOther) { return new Vector(argSelf._fieldX-argOther._fieldX, argSelf._fieldY-argOther._fieldY, argSelf._fieldZ-argOther._fieldZ); }
    public Vector GetNormal()
    {
    	Comparinator varComp = new Comparinator();
    	double varMagnitude = GetMagnitude();
    	return varComp.CheckFloat(varMagnitude, 0) ? new Vector(0.0,0.0,0.0) : new Vector(_fieldX/varMagnitude, _fieldY/varMagnitude, _fieldZ/varMagnitude);
    }
    public Vector GetCrossProduct(Vector argOther)
    {
    	return new Vector(_fieldY * argOther._fieldZ - _fieldZ * argOther._fieldY, _fieldZ * argOther._fieldX - _fieldX * argOther._fieldZ, _fieldX * argOther._fieldY - _fieldY * argOther._fieldX);
    }
    public Vector GetReflect(Vector argNormal) { return argNormal * 2.0 * GetDotProduct(argNormal); }
}
